### ISSUE TITLE
Structured (Flat) JSON Log Format

### ISSUE THREAD

**Issue by t3chn0m4g3** [mushorg/glutton#110]
Glutton already supports JSON logging in a flat structure. This is really great, as it allows for easily ingesting logs by Logstash.

However, at the moment the logging is not structured, important information i.e. src_ip, dest_ip, src_port, dest_port, etc. is just written to the `msg` field and makes further processing in the ELK stack (Kibana visualizations, etc.) impossible.

Would it be possible to support a structured yet flat JSON format (ELK does not like nested fields) with its unique fields for src_ip, dest_ip, src_port, dest_port, payload, etc.?

**Comment by glaslos** [comment]
I just a small change to master. Can you tell me if it's going into the right direction?

**Comment by t3chn0m4g3** [comment]
Thank you, this is going into right direction. A sample output how it will ingested by logstash / kibana:

![image](https://user-images.githubusercontent.com/4318452/40062399-6d10b6a6-585b-11e8-9dec-dd6b6c209cb3.png)
```
{
  "_index": "logstash-2018.05.15",
  "_type": "doc",
  "_id": "h3gTZGMBmEcc656Fn2qq",
  "_version": 1,
  "_score": null,
  "_source": {
    "sensorID": "e79f21b1-69f1-40bd-b78c-00f3f13d3871",
    "level": "info",
    "host": "57cc74bc23cd",
    "caller": "glutton/tcp.go:25",
    "path": "/data/glutton/log/glutton.log",
    "@version": "1",
    "msg": "[log.tcp ] 172.20.254.130\n00000000  16 03 01 02 00 01 00 01  fc 03 03 41 49 46 b6 46  |...........AIF.F|\n00000010  56 78 5e 0b 12 3c ec 61  ba 20 31 c9 79 8a 3e 0a  |Vx^..<.a. 1.y.>.|\n00000020  21 e3 79 27 43 8b 63 f8  81 45 ea 20 9d 17 a2 5a  |!.y'C.c..E. ...Z|\n00000030  6d 89 6e 66 f3 51 33 20  d6 44 02 48 c7 18 76 b8  |m.nf.Q3 .D.H..v.|\n00000040  4f 04 28 8d 7c 46 89 cb  08 fe f3 b1 00 22 5a 5a  |O.(.|F.......\"ZZ|\n00000050  13 01 13 02 13 03 c0 2b  c0 2f c0 2c c0 30 cc a9  |.......+./.,.0..|\n00000060  cc a8 c0 13 c0 14 00 9c  00 9d 00 2f 00 35 00 0a  |.........../.5..|\n00000070  01 00 01 91 da da 00 00  ff 01 00 01 00 00 17 00  |................|\n00000080  00 00 23 00 00 00 0d 00  14 00 12 04 03 08 04 04  |..#.............|\n00000090  01 05 03 08 05 05 01 08  06 06 01 02 01 00 05 00  |................|\n000000a0  05 01 00 00 00 00 00 12  00 00 00 10 00 0e 00 0c  |................|\n000000b0  02 68 32 08 68 74 74 70  2f 31 2e 31 75 50 00 00  |.h2.http/1.1uP..|\n000000c0  00 0b 00 02 01 00 00 33  00 2b 00 29 fa fa 00 01  |.......3.+.)....|\n000000d0  00 00 1d 00 20 0c 42 69  5a 71 ba 5c 93 f9 f1 01  |.... .BiZq.\\....|\n000000e0  98 11 79 dc fc c0 61 dd  29 db af 49 73 2c a1 66  |..y...a.)..Is,.f|\n000000f0  24 a7 cd 99 55 00 2d 00  02 01 01 00 2b 00 0b 0a  |$...U.-.....+...|\n00000100  aa aa 7f 17 03 03 03 02  03 01 00 0a 00 0a 00 08  |................|\n00000110  fa fa 00 1d 00 17 00 18  ca ca 00 01 00 00 15 00  |................|\n00000120  e4 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000140  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000150  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000160  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000170  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000180  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000190  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n000001a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n000001b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n000001c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n000001d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n000001e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n000001f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|\n00000200  00 00 00 00 00                                    |.....|\n",
    "@timestamp": "2018-05-15T13:53:16.733Z",
    "type": "Glutton"
  },
  "fields": {
    "@timestamp": [
      "2018-05-15T13:53:16.733Z"
    ]
  },
  "highlight": {
    "type": [
      "@kibana-highlighted-field@Glutton@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1526392396733
  ]
}
```

Another example for telnet:
![image](https://user-images.githubusercontent.com/4318452/40062912-824bae8a-585c-11e8-8a0c-89f8aa8cba0e.png)

```
{
  "_index": "logstash-2018.05.15",
  "_type": "doc",
  "_id": "tngYZGMBmEcc656FTGo6",
  "_version": 1,
  "_score": null,
  "_source": {
    "sensorID": "1f486167-0caf-4f34-9d84-ad66994abe2d",
    "level": "info",
    "host": "57cc74bc23cd",
    "caller": "glutton/telnet.go:63",
    "path": "/data/glutton/log/glutton.log",
    "@version": "1",
    "msg": "[telnet  ] send: \"Password: \"",
    "@timestamp": "2018-05-15T13:58:23.300Z",
    "type": "Glutton"
  },
  "fields": {
    "@timestamp": [
      "2018-05-15T13:58:23.300Z"
    ]
  },
  "highlight": {
    "type": [
      "@kibana-highlighted-field@Glutton@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1526392703300
  ]
}
```

In both examples you can see new index fields introduced (yellow yield icon) which basically means ELK can now index the contents of these fields. It would be ideal, if there were the following fields:
- `src_ip`
- `dest_ip`
- `src_port`
- `dest_port`
- `proto`
- `payload` (currently `msg`)
- `payload_type` (send or receive)
- rename `msg` to `message`

... and if there is still room for whip-cream and cherry on top ... analyzed payload fields i.e.
- `username`
- `password`

Here an example of Heralding where i.e. src_ip, dest_ip, etc. are being indexed:
![image](https://user-images.githubusercontent.com/4318452/40063232-25d5d8b4-585d-11e8-9af5-5ad72fc2b2c7.png)

```
{
  "_index": "logstash-2018.05.15",
  "_type": "doc",
  "_id": "cHgRZGMBmEcc656Fq2re",
  "_version": 1,
  "_score": null,
  "_source": {
    "t-pot_ip_ext": "87.150.238.182",
    "geoip": {},
    "@version": "1",
    "type": "Heralding",
    "password": "test",
    "t-pot_ip_int": "172.20.254.120",
    "username": "test",
    "proto": "telnet",
    "message": "2018-05-15 13:51:09.344991,868083a7-9d84-475e-8411-68f320b0a24e,b23fa459-1046-48a3-82b0-2058b2d0c0f8,172.20.0.1,38318,84.154.40.73,23,telnet,test,test\r",
    "host": "57cc74bc23cd",
    "path": "/data/heralding/log/auth.csv",
    "src_ip": "172.20.0.1",
    "dest_ip": "84.154.40.73",
    "@timestamp": "2018-05-15T13:51:09.344Z",
    "src_port": 38318,
    "auth_id": "868083a7-9d84-475e-8411-68f320b0a24e",
    "tags": [
      "_geoip_lookup_failure"
    ],
    "t-pot_hostname": "spottycathedral",
    "session_id": "b23fa459-1046-48a3-82b0-2058b2d0c0f8",
    "dest_port": 23
  },
  "fields": {
    "@timestamp": [
      "2018-05-15T13:51:09.344Z"
    ]
  },
  "highlight": {
    "type": [
      "@kibana-highlighted-field@Heralding@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1526392269344
  ]
}
```
In case of questions feel free to PM me.

**Comment by nassimabedi** [comment]
@glaslos @t3chn0m4g3 Thank you :)
I haven't worked with ELK but I think something like bot detection.
Can I use patterns in a sequence way to search and detect bots in ELK for example mirai

**Comment by t3chn0m4g3** [comment]
Fixed by 7b4e2be2881c0b97cf251ca4199549f32ca71835

@glaslos Thank you 😄
